woocommerce/woocommerce-google-analytics-integration
# Separate the global site tag setup and event tracking

### Changes proposed in this Pull Request:

The changes to how the tracker is initiated in version `2.0.0` has caused problems with more sites than anticipated. Primarily those problems come from plugins that minify and combine JS files as many of them disregard listed dependencies while doing so.

As it currently stands there are many optimisation plugins that do work well with Google Analytics for WooCommerce but to make the extension more resilient against unexpected scenarios from different optimisation techniques this PR restructures how things execute.

- `dataLayer` and `gtag` (_or a function using the filtered tracker var name_) are no longer included in the JS file. Instead they are added directly to the page `<head>`
    - This is closer to how it was functioning pre-2.0.0
    - Does not use the WordPress script queue system to prevent this block of code being moved to anywhere else
    - Hooks into `wp_head` early so that the script is added as soon as possible. This makes it accessible to other plugins and means we are able to rely on it being available on the page regardless of when our scripts load 582e41b
- Relies on the `DOMContentLoaded` event to ensure inline script data is available at the point we begin tracking events 
    - E2E tests have been added to cover some of the scenarios we've seen with optimisation plugins that has caused problems with this 8e7481d
- Restructures the tracker to be more functional 9fede11

### Detailed test instructions:

1. Build extension from `update/separate-site-tag-setup-and-event-tracking`
2. Run [smoke tests](https://github.com/woocommerce/woocommerce-google-analytics-integration/wiki/General-Testing#for-compatibility-and-general-smoke-testing) to confirm all events are tracked
3. Install an optimisation plugin like [Autoptimize](https://en-gb.wordpress.org/plugins/autoptimize/)
    - Configure the extension to combine and minify all Javascript
4. Check the page source to confirm that `gtag` is still setup in the page `<head>`
5. Re-run smoke tests to confirm events are still tracked

### Additional details:

- An adjustment was made to the selector for the related products add to cart E2E test as it was occasionally failing if a variable product was the first displayed in the related products list 25bd09c

### Changelog entry

> Update - Separate the site tag from the event tracking file and delay execution until DOMContentLoaded
